Userland/Applications/SoundPlayer/M3UParser.cpp

```diff
@@ -0,0 +1,130 @@
+}
+
+NonnullOwnPtr<Vector<M3UEntry>> M3UParser::parse(bool include_extended_info)
+{
+    auto vec = make<Vector<M3UEntry>>();
+
+    bool has_exteded_info_tag = false;
+    if (!m_use_utf8) {
+        auto lines = m_m3u_raw_data.split_view('\n');
+
+        if (include_extended_info) {
+            if (lines[0] == "#EXTM3U")
+                has_exteded_info_tag = true;
+        }
+
+        M3UExtendedInfo metadata_for_next_file {};
+        for (auto& line : lines) {
+            line = line.trim_whitespace();
+            M3UEntry entry {};
+            if (line.starts_with('#') && has_exteded_info_tag) {
+                if (line.starts_with("#EXTINF")) {
+                    auto data = line.substring_view(8);
+                    auto separator = data.find_first_of(',');
+                    VERIFY(separator.has_value());
+                    auto seconds = data.substring_view(0, separator.value());
+                    VERIFY(!seconds.is_whitespace() && !seconds.is_null() && !seconds.is_empty());
+                    metadata_for_next_file.track_length_in_seconds = seconds.to_uint();
+                    auto display_name = data.substring_view(seconds.length() + 1);
+                    VERIFY(!display_name.is_empty() && !display_name.is_null() && !display_name.is_empty());
+                    metadata_for_next_file.track_display_title = display_name;
+                    //TODO: support the alternative, non-standard #EXTINF value of a key=value dictionary
+                } else if (line.starts_with("#PLAYLIST")) {
+                    auto name = line.substring_view(10);
+                    VERIFY(!name.is_empty());
+                    m_parsed_playlist_title = name;
+                } else if (line.starts_with("#EXTGRP")) {
+                    auto name = line.substring_view(8);
+                    VERIFY(!name.is_empty());
+                    metadata_for_next_file.group_name = name;
+                } else if (line.starts_with("#EXTALB")) {
+                    auto name = line.substring_view(8);
+                    VERIFY(!name.is_empty());
+                    metadata_for_next_file.album_title = name;
+                } else if (line.starts_with("#EXTART")) {
+                    auto name = line.substring_view(8);
+                    VERIFY(!name.is_empty());
+                    metadata_for_next_file.album_artist = name;
+                } else if (line.starts_with("#EXTGENRE")) {
+                    auto name = line.substring_view(10);
+                    VERIFY(!name.is_empty());
+                    metadata_for_next_file.album_genre = name;
+                }
+                //TODO: Support M3A files (M3U files with embedded mp3 files)
+            } else {
+                entry.path = line;
+                entry.extended_info = metadata_for_next_file;
+                vec->append(entry);
+            }
+        }
+    } else {
+        //TODO: Implement M3U8 parsing
+        VERIFY_NOT_REACHED();
```

Comment:
Nit: this could be a [`TODO();`](https://github.com/SerenityOS/serenity/blob/master/AK/Assertions.h#L36) instead of a `VERIFY_NOT_REACHED();`. It seems more contextually appropriate :^)

Reply:
That's a nice alias for contextual info :+1: